apps/buckler_bot/lib/buckler_bot/handler.ex

```diff
@@ -114,15 +95,47 @@ defmodule BucklerBot.Handler do
       }
     }
   }, _) do
-    case Repo.user_unauthorized?(chat_id, user_id) do
+    Logger.debug("New media message!")
+    case Connections.user_unauthorized?(chat_id, user_id) do
       {true, user} ->
-        Repo.delete_user(chat_id, user_id)
+        process_captcha_check(false, conn, message_id, user)
+      _ -> conn |> Agala.Conn.halt
+    end
+  end
+
+  def process_captcha_check(true, conn, message_id, user) do
+    multi do
+      add delete_message(conn, user.chat_id, user.welcome_message_id)
+      add delete_message(conn, user.chat_id, message_id)
+      Connections.delete_chatuser(user.chat_id, user.user_id)
+    end
+  end
+  def process_captcha_check(false, conn, message_id, user) do
+    case user.attempts < 2 do
```

Comment:
You decrease attempts, so user ready to be banned if user.attempts == 0 ?

So maybe this solution is cleaner? 
```
def process_captcha_check(false, conn, message_id, %Customer{attempts: 0}=user) do # ban!
  multi do
    {:ok, user} = Connections.delete_chatuser(user.chat_id, user.user_id)
    add delete_message(conn, user.chat_id, user.welcome_message_id)
    ...
  end
end

def process_captcha_check(false, conn, message_id, %Customer{attempts: attempts}=user) do # decrease attempt
  with %{captcha: captcha, answer: answer} <- BucklerBot.Captcha.generate_captcha(user.lang),
        {:ok, user} <- Connections.decrease_attempts(user.chat_id, user.user_id, answer)
        ...
  end
end
```

Reply:
No. If something went wrong , for example server will crash after decreasing attempts to 0, the process will continue infinity because number of attempts will be -1,-2 and so on

Reply:
Hm, you can use guards
```
def process_captcha_check(false, conn, message_id, %Customer{attempts: attempts}=user) when attempts <= 0 do # ban!
  multi do
    {:ok, user} = Connections.delete_chatuser(user.chat_id, user.user_id)
    add delete_message(conn, user.chat_id, user.welcome_message_id)
    ...
  end
end

def process_captcha_check(false, conn, message_id, %Customer{attempts: attempts}=user) do # decrease attempt
  with %{captcha: captcha, answer: answer} <- BucklerBot.Captcha.generate_captcha(user.lang),
        {:ok, user} <- Connections.decrease_attempts(user.chat_id, user.user_id, answer)
        ...
  end
end
```